File: app-flutter/android/app/src/main/kotlin/de/haphone/app/test/ring/Appearance.kt
package de.haphone.app.test.ring

import android.content.Context
import android.content.res.Configuration
import android.util.Log

/**
 * In-app "Erscheinungsbild" (Ich tab): Dunkel, Hell or Wie System. Nachtwache
 * is dark-first, so anything unknown or missing means [DARK]. The wire values
 * are the ones Dart sends via `setAppearance` (lib/services/appearance.dart).
 */
enum class Appearance(val wire: String) {
    DARK("dark"),
    LIGHT("light"),
    SYSTEM("system");

    /** Effective dark theme for this choice; [systemNight] only matters for [SYSTEM]. */
    fun isDark(systemNight: Boolean): Boolean = when (this) {
        DARK -> true
        LIGHT -> false
        SYSTEM -> systemNight
    }

    companion object {
        val DEFAULT = DARK

        fun fromWire(value: String?): Appearance = entries.firstOrNull { it.wire == value } ?: DEFAULT

        /** Night flag of a Configuration.uiMode value. */
        fun isNightUiMode(uiMode: Int): Boolean =
            (uiMode and Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_YES
    }
}

/**
 * Persists [Appearance] natively so the ringing screen (IncomingCallActivity)
 * can apply it even when it starts without the Flutter engine running.
 */
object AppearanceStore {
    private const val TAG = "Appearance"
    private const val PREFS = "haphone_appearance"
    private const val KEY_MODE = "mode"

    fun load(ctx: Context): Appearance =
        Appearance.fromWire(ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getString(KEY_MODE, null))

    fun save(ctx: Context, appearance: Appearance) {
        ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().putString(KEY_MODE, appearance.wire).apply()
        Log.i(TAG, "saved ${appearance.wire}")
    }

    /** Effective dark theme for [ctx] (stored choice + the context's current night mode). */
    fun isDark(ctx: Context): Boolean =
        load(ctx).isDark(Appearance.isNightUiMode(ctx.resources.configuration.uiMode))
}

File: app-flutter/lib/services/appearance.dart
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'local_store.dart';
import 'sip_channel.dart';

/// In-app "Erscheinungsbild" (Ich tab). Nachtwache is dark-first, so the
/// default is [dark]; [system] follows the phone. [wire] is what the native
/// side stores (ring/Appearance.kt).
enum AppAppearance {
  dark('dark', 'Dunkel', ThemeMode.dark),
  light('light', 'Hell', ThemeMode.light),
  system('system', 'Wie System', ThemeMode.system);

  const AppAppearance(this.wire, this.label, this.themeMode);

  final String wire;
  final String label;
  final ThemeMode themeMode;

  static const fallback = AppAppearance.dark;

  static AppAppearance fromWire(String? value) =>
      values.firstWhere((a) => a.wire == value, orElse: () => fallback);
}

/// Holds the appearance, persists it (SharedPreferences) and pushes it to the
/// native side. MaterialApp listens and switches live.
class AppearanceController extends ValueNotifier<AppAppearance> {
  AppearanceController({
    Future<SharedPreferences?> Function()? prefs,
    Future<void> Function(String mode)? pushNative,
  })  : _prefs = prefs ?? loadPrefs,
        _pushNative = pushNative ?? SipChannel.instance.setAppearance,
        super(AppAppearance.fallback);

  static final AppearanceController instance = AppearanceController();

  final Future<SharedPreferences?> Function() _prefs;
  final Future<void> Function(String mode) _pushNative;

  ThemeMode get themeMode => value.themeMode;

  /// Reads the stored choice and pushes it to native once (app start).
  Future<void> load() async {
    final prefs = await _prefs();
    value = AppAppearance.fromWire(prefs?.getString(StoreKeys.appearance));
    await _push();
  }

  Future<void> set(AppAppearance appearance) async {
    value = appearance;
    final prefs = await _prefs();
    await prefs?.setString(StoreKeys.appearance, appearance.wire);
    await _push();
  }

  Future<void> _push() async {
    try {
      await _pushNative(value.wire);
    } catch (e) {
      debugPrint('setAppearance failed: $e');
    }
  }
}
